File: app/src/main/java/ua/com/foxminded/locationtrackera/model/usecase/SendLocationsUseCaseImpl.java
package ua.com.foxminded.locationtrackera.model.usecase;

import io.reactivex.rxjava3.core.Single;

import ua.com.foxminded.locationtrackera.model.locations.LocationRepository;
import ua.com.foxminded.locationtrackera.util.Result;

public class SendLocationsUseCaseImpl implements SendLocationsUseCase {

    private final LocationRepository repository;

    public SendLocationsUseCaseImpl(LocationRepository repository) {
        this.repository = repository;
    }

    @Override
    public Single<Result<Void>> execute() {
        return repository.sendLocations(repository.getAllLocations())
                .onErrorResumeNext(e -> Single.just(new Result.Error(e)));
    }
}

File: models_impl/src/main/java/ua/com/foxminded/locationtrackera/models_impl/usecase/SendLocationsUseCaseImpl.kt
package ua.com.foxminded.locationtrackera.models_impl.usecase

import io.reactivex.rxjava3.core.Single
import ua.com.foxminded.locationtrackera.models.locations.UserLocation
import ua.com.foxminded.locationtrackera.models.usecase.SendLocationsUseCase
import ua.com.foxminded.locationtrackera.models.util.Result

class SendLocationsUseCaseImpl(private val repository: ua.com.foxminded.locationtrackera.models.locations.LocationRepository) :
    SendLocationsUseCase {

    override fun execute(): Single<Result<Void?>> =
        Single.fromCallable { repository.getAllLocations() }
            .flatMap { locationList: List<UserLocation> -> repository.sendLocations(locationList) }
            .onErrorResumeNext { e: Throwable? -> Single.just(Result.Error(e)) }
}
